### DISCUSSION TITLE
PostgresOperator fails with AirflowException: Unknown hook type "postgresql+psycopg2"

### DISCUSSION THREAD

**Discussion by kiranzo** [apache/airflow#34821] [Q&A] [answered]
What it says in the title.
I defined target database connection in my docker-compose file like this:

`AIRFLOW_CONN_TARGETDB: "postgresql+psycopg2://admin:password@127.0.0.1:5432/TARGETDATABASE"`

I already used this connection with PostgresHook, and it worked as it should, even from a test python script with the same python instance as airflow:

`hook = PostgresHook(postgres_conn_id="TARGETDB")`

But when I try to use the same postgres_conn_id for PostgresOperator, I get

```
 ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/providers/common/sql/operators/sql.py", line 263, in execute
    hook = self.get_db_hook()
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/providers/common/sql/operators/sql.py", line 181, in get_db_hook
    return self._hook
  File "/usr/local/lib/python3.10/functools.py", line 981, in __get__
    val = self.func(instance)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/providers/common/sql/operators/sql.py", line 144, in _hook
    hook = conn.get_hook(hook_params=self.hook_params)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/connection.py", line 330, in get_hook
    raise AirflowException(f'Unknown hook type "{self.conn_type}"')
airflow.exceptions.AirflowException: Unknown hook type "postgresql+psycopg2"
```

My code for PostgresOperator looks like this:

```
from airflow.providers.postgres.operators.postgres import PostgresOperator
.......

set_db_timestamp_format = PostgresOperator(
        task_id="set_db_timestamp_format", 
        postgres_conn_id="TARGETDB", 
        sql=SET_DATETIME_FORMAT_SQL)

```

What am I doing wrong? Was I supposed to install some extra libraries to work with PostgresOperator?

**Comment by Taragolis** [comment]
Airflow Connection URI it is not same as SQLAlchemy URL or Psycopg2 DSN, so you need to use `postgres` instead of `postgresql+psycopg2`

**Reply by kiranzo** [reply]
How do people go around this normally? Give up on using PostgresOperator and only use PythonOperator with PostgresHook? 
Don't tell me they make two different connections to the same database for two different operators. :(

**Reply by Taragolis** [reply, marked as answer]
It might work with hook, because it ignore connection type when you directly pass this connection id to the Hook, let's say it work accidentally.

The things are different when you use it within operator it use generic SqlOperator which determine correct hook by associated connection type, and `postgresql+psycopg2` it is unknown connection type that is why you have this error. Just use for [Airflow PostgreSQL Connection](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/connections/postgres.html) `postgres` as connection type

**Reply by kiranzo** [reply]
This works, thanks!
